Add binary sensors to Volvo integration

thomasddn opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add binary sensors to Volvo integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40449
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

thomasddn committed:
Add binary sensors

NoRi2909 commented on homeassistant/components/volvo/strings.json:
```diff
@@ -1,4 +1,10 @@
 {
+  "common": {
+    "high_pressure": "High pressure",
+    "low_pressure": "Low pressure",
+    "pressure": "Pressure",
+    "very_low_pressure": "Very low pressure"
```
That duplicates "Pressure" on the state lines, for example:

    Pressure         Very low pressure

If I translate that into German we end up with

    Luftdruck        Sehr niedriger Luftdruck

That will probably no longer fit the default column width on desktop.


Instead of redefining the states for "High", "Low" and "Very low" here you could simply use the common states instead for all four tires below.

```
"[%key:common::state::high%]"
"[%key:common::state::low%]"
"[%key:common::state::very_low%]"
```

Then the German translation becomes

    Luftdruck         Sehr niedrig

CFenner commented on homeassistant/components/volvo/binary_sensor.py:
```diff
@@ -0,0 +1,416 @@
+        key="tailgate",
+        api_field="tailgate",
+    ),
+    VolvoCarsDoorDescription(
+        key="tank_lid",
+        api_field="tankLid",
+    ),
```
Is there a sensor for the charging lid as well?

thomasddn replied:
No, it's not available in the API

CFenner commented on homeassistant/components/volvo/binary_sensor.py:
```diff
@@ -0,0 +1,416 @@
+    ),
+    # warnings endpoint
+    VolvoBinarySensorDescription(
+        key="fog_light_front_warning",
```
I wonder if this kind of detail is too much, having a sensor for every light in the car. 
Maybe it would be more suitable to have a *lights warning sensor* that states in the attributes which light is affected?

thomasddn replied:
I think that will be too much to put in one sensor. Also see comment https://github.com/home-assistant/core/pull/150127#discussion_r2260164209

CFenner commented on homeassistant/components/volvo/binary_sensor.py:
```diff
@@ -0,0 +1,416 @@
+        key="window_rear_right",
+        api_field="rearRightWindow",
+    ),
+    VolvoCarsWindowDescription(
+        key="sunroof",
+        api_field="sunroof",
+    ),
```
Is there a sensor for the roof window shield as well?

thomasddn replied:
The API only mentions `sunroof`. I don't have any of these, so I can't tell for sure if the API makes a distinction between the two.

CFenner commented on homeassistant/components/volvo/binary_sensor.py:
```diff
@@ -0,0 +1,416 @@
+        on_values=("TOO_LOW",),
+    ),
+    # engine-status endpoint
+    VolvoBinarySensorDescription(
+        key="engine_status",
+        api_field="engineStatus",
+        device_class=BinarySensorDeviceClass.RUNNING,
+        on_values=("RUNNING",),
+    ),
```
I'm not a big fan of the `RUNNING` device class, in German it translates to "out of order" (which may also be a translation issue), I would rather prefer "on" / "off", especially as I cannot change the sensor appearance to use the default again.

<img width="299" height="33" alt="Bildschirmfoto 2025-08-06 um 22 53 27" src="https://github.com/user-attachments/assets/5c9a4bbd-7ca6-4037-bd01-f8565279a90c" />

NoRi2909 replied:
We have that German translation for pumps as well (like in ViCare), there "In Betrieb" / "Außer Betrieb" for "Running" / "Not running" of a circulation pump works quite OK.

The translation has historical reasons in the German translation as "running" was first understood as applying to a program and translated as "Wird ausgeführt"  / "Wird nicht ausgeführt" which was even more off for a pump or an engine.

We can see if we can change all the states, triggers and conditions to "Läuft" / Läuft nicht" but I have to check if this does not create new stylistic issues.

thomasddn replied:
I'll resolve this comment then.

CFenner commented on homeassistant/components/volvo/icons.json:
```diff
@@ -1,5 +1,271 @@
+          "on": "mdi:car-door-lock-open"
+        }
+      },
+      "tank_lid": {
+        "default": "mdi:car-door-lock",
+        "state": {
+          "on": "mdi:car-door-lock-open"
+        }
+      },
```
There are a lot of sensors with a *door* icon. Would be nice to find some more suitable icons, but I guess you already did a search.

thomasddn committed:
Simplify tire attributes

joostlek commented on homeassistant/components/volvo/binary_sensor.py:
```diff
@@ -0,0 +1,416 @@
+) -> None:
+    """Set up binary sensors."""
+    coordinators = entry.runtime_data
+    entities = [
+        VolvoBinarySensor(coordinator, description)
+        for coordinator in coordinators
+        for description in _DESCRIPTIONS
+        if description.api_field in coordinator.data
+    ]
+
+    async_add_entities(entities)
```
```suggestion
    async_add_entities(
        VolvoBinarySensor(coordinator, description)
        for coordinator in coordinators
        for description in _DESCRIPTIONS
        if description.api_field in coordinator.data
    )
```

joostlek commented on homeassistant/components/volvo/binary_sensor.py:
```diff
@@ -0,0 +1,416 @@
+            else None
+        )
+
+        if self.entity_description.api_value_in_attributes:
+            attribute_value = (
+                api_field.value
+                if self.entity_description.api_value_attribute_fn is None
+                else self.entity_description.api_value_attribute_fn(api_field.value)
+            )
+
+            if isinstance(attribute_value, str):
+                attribute_value = value_to_translation_key(attribute_value)
+
+            self._attr_extra_state_attributes[
+                self.entity_description.api_value_attribute_name
+            ] = attribute_value
```
Generally we don't like extra state attributes, can you add what kind of attributes we add here? (I tried to understand it, but it's probably quicker to ask)

joostlek replied:
Looking at the strings.json I see oil level, tire pressure, etc, I think they can be their own enum sensor and maybe disabled by default

gjohansson-ST replied:
It seems we didn't remove the attributes?

thomasddn committed:
Code style

thomasddn committed:
Merge remote-tracking branch 'upstream/dev' into volvo/binary-sensor

CFenner commented:
Sensors are still available when connection is terminated.

<img width="338" height="507" alt="Bildschirmfoto 2025-08-11 um 14 48 48" src="https://github.com/user-attachments/assets/e87cbced-01ef-40d6-80f7-e2ffcda63b8e" />

thomasddn committed:
Merge remote-tracking branch 'upstream/dev' into volvo/binary-sensor

thomasddn committed:
Add disabled sensors

thomasddn committed:
Merge remote-tracking branch 'upstream/dev' into volvo/binary-sensor

thomasddn committed:
Sort descriptions by key

thomasddn committed:
Simplify adding sensors

gjohansson-ST commented:
Maybe we can do the additions in `sensor.py` in a preliminary PR? It's quite extensive this PR.

thomasddn commented:
@gjohansson-ST The additions to `sensor.py` were made upon feedback in this PR. Pulling apart all changes will be quite some work as the additions for `sensor.py` are based on code to support binary sensors.

I understand the question and acknowledge it's a bit of a big(ger than expected) PR, but it had its reasons.

There are two parts in this PR, maybe it can be reviewed this way:

1. Added support for binary sensors
2. Added additional sensors

If you insist, I can add the second part in a new PR, based on this PR, but I prefer to not create a "preliminary" one.

thomasddn commented:
@gjohansson-ST Yes, that way we stay under 5000 requests per day and can add another car with the same API key.

thomasddn committed:
Revert "Add disabled sensors"

This reverts commit 07ccde139668fc71a5f55b3b772f2214ce4e580f.

thomasddn committed:
Sort descriptions

thomasddn committed:
Clarify use of multiple coordinators

thomasddn committed:
Mark some binary sensors as diagnostic

thomasddn committed:
Update snapshot

thomasddn committed:
Merge remote-tracking branch 'upstream/dev' into volvo/binary-sensor

thomasddn committed:
Remove attributes & service warning

thomasddn committed:
Remove service warning icon
